flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/sink/CachingTableSupplier.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.sink;
+
+import java.time.Duration;
+import org.apache.flink.util.Preconditions;
+import org.apache.iceberg.SerializableTable;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.util.DateTimeUtil;
+import org.apache.iceberg.util.SerializableSupplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A table loader that will only reload a table after a certain interval has passed. WARNING: This
+ * table loader should be used carefully when used with writer tasks. It could result in heavy load
+ * on a catalog for jobs with many writers.
+ */
+class CachingTableSupplier implements SerializableSupplier<Table> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CachingTableSupplier.class);
+
+  private final Table initialTable;
+  private final TableLoader tableLoader;
+  private final Duration tableRefreshInterval;
+  private long nextReloadTimeMs;
+  private transient Table table;
+
+  CachingTableSupplier(
+      SerializableTable initialTable, TableLoader tableLoader, Duration tableRefreshInterval) {
+    Preconditions.checkArgument(initialTable != null, "initialTable cannot be null");
+    Preconditions.checkArgument(tableLoader != null, "tableLoader cannot be null");
+    Preconditions.checkArgument(
+        tableRefreshInterval != null, "tableRefreshInterval cannot be null");
+    this.initialTable = initialTable;
+    this.table = initialTable;
+    this.tableLoader = tableLoader;
+    this.tableRefreshInterval = tableRefreshInterval;
+    this.nextReloadTimeMs = System.currentTimeMillis() + tableRefreshInterval.toMillis();
+  }
+
+  @Override
+  public Table get() {
+    if (table == null) {
+      this.table = initialTable;
```

Comment:
wondering if stale `initialTable` object can be used after restoring from checkpoint? is it correct to actually store the `initialTable`? should it always load the table on first get?

Reply:
The table will be refreshed in `IcebergStreamWriter.open()` after the task writer is initialized. Keeping the initial table is a temporary measure, so the initial table schema and partition spec are used by the task writer after a restart, until schema evolution is supported.

Reply:
yes, I also checked the refresh call on open method. but refresh always check the time threshold. so before the time threshold elapsed, stale `initialTable` is use, right?

Reply:
The time of the initial table load is also serialized, so if it is stale then the refresh will update it.

Reply:
I changed this slightly, so the last load time is stored instead of the next load time, which should be a little bit more clear.

Reply:
>The time of the initial table load is also serialized, so if it is stale then the refresh will update it.

ah. I missed this point earlier.